### PR TITLE
feat: sort case table items by attribute [PT-181890013]

### PR DESCRIPTION
**Pull request by kswenson**
[[PT-181890013]](https://www.pivotaltracker.com/story/show/181890013)

- Required some interesting details including a custom undo/redo implementation.
- Adds `caseIdsHash` and `caseIdsOrderedHash` to the `CollectionModel`.
- Adds `hashOrderedStringSet` function to `js-utils`.
- Code for comparing values largely ported from v2.

### CODE REVIEW

File: v3/src/utilities/data-utils.ts [hunk 1]
```diff
@@ -42,3 +43,67 @@ export const numericSortComparator = function ({a, b, order}: ICompareProps): nu
...
+    /* istanbul ignore next */
+    default: return kTypeUnknown
+  }
+}
+
+export function sortableValue(value: any) {
+  const type = typeCode(value)
+  let num = type === kTypeNumber ? value : NaN
+  // strings convertible to numbers are treated numerically
+  if (type === kTypeString && value.length) {
```
Comment (scytacki) [inline v3/src/utilities/data-utils.ts:76]: Shouldn't this check if the attribute has a manually set type? For example if the attribute type is string then I'd expect sorting `["001","001a"]` to treat the `"001"` as a string. If I'm reading the logic right here that array would end up reversed without taking into account the attribute type.

File: v3/src/utilities/js-utils.ts [hunk 2]
```diff
@@ -173,3 +173,15 @@ export function hashStringSets(stringSets: Array<string[]>) {
...
     .reduce((acc, hash) => acc ^ hash, 0) // XOR all individual hashes
 }
+
+/*
+ * hashOrderedStringSet()
+ *
+ * returns an order-dependent hash value for a set of strings (e.g. ids).
+ * developed with the help of ChatGPT.
+ */
+export function hashOrderedStringSet(strings: string[]) {
```
Comment (scytacki) [inline v3/src/utilities/js-utils.ts:183]: If we find this isn't working well (speed or accuracy) we could try doing `hashString(JSON.stringify(strings))`

**Review by scytacki — APPROVED** [review]
This looks good. Perhaps it should be expanded to take into account an attribute type when doing the sort. But I'd guess that is not something v2 does, so it should probably be a different story/PR.

**Comment by kswenson** [comment]
> This looks good. Perhaps it should be expanded to take into account an attribute type when doing the sort. But I'd guess that is not something v2 does, so it should probably be a different story/PR.

Interesting thought, but this is the way v2 handled it and changing it would require a larger conversation.

File: v3/src/utilities/js-utils.ts (hunk 2 above)
Comment (kswenson) [reply v3/src/utilities/js-utils.ts:183]: I would expect the `JSON.stringify` version to be slower, but it'd be interesting to check.

**Pull request merged**